(XMB) entry enumeration

DisasterMo opened the pull request:
## Description
This PR adds a string to the bottom right corner of the screen in XMB depicting _entry index/number of entries in current list_ :
![entry enumeration xmb](https://user-images.githubusercontent.com/49207524/119189348-7c9f1700-ba6b-11eb-8e7e-b7ab6cb862e8.png)

It is marked WIP because I would like to hear some opinions on the following concern I have:
The lists in XMB, be it the games, the settings or even the quick menu, all look, and for the most part behave, kinda the same. In addition to that, not every list that seems like it should be a 'dedicated' playlist, is (Images is a playlist, but Music and Videos aren't).
Therefore, I think it would make sense to utilise this entry enumeration for all lists in XMB, not just the playlists.

I have implemented an option (Settings -> Playlists -> Show Entry Index For All Lists), which will toggle between enumerating the entries for all lists and only enumerating playlist entries. So please, try it out and let me know your preferences and thoughts on how to handle this.

Thank you.

## Related Pull Requests
This is a continuation of #12398.

## Reviewers
A big thank you to @jdgleaver for providing a great example for creating settings entries.

gouchi commented:
Thank you for this new feature. I did some "monkey tests" and  it is working.

>  Therefore, I think it would make sense to utilise this entry enumeration for all lists in XMB, not just the playlists.

I was a little confused as it was in every list not only playlists.

jdgleaver commented:
@DisasterMo I think this looks really nice! I have a few comments, though :)

- I agree with gouchi. The index display does seem out of place on non-playlist-type menus. I would suggest removing the `SHOW_ALL_IDX` setting, and instead just apply the index to the following: normal playlists, history, music, video, images, and I reckon it would fit well with the 'load content' file browser too

- The index display in its current location will overlap thumbnails in many cases. There are a couple of way you could deal with this: (a) when the index is enabled, reduce the available thumbnail height a little or (b) when thumbnail vertical disposition is enabled, move the index display to the left hand side of the screen, maybe inline with the currently selected entry. No sure which would look best...

- `snprintf()` is quite slow. Rather than calling it every frame, it would be good to cache the result when the selection index changes

- The `char path[PATH_MAX_LENGTH + 286];` change in `task_save.c` should be omitted from this PR. It's unrelated to the added functionality here, and it's better to silence the warning by checking (then ignoring) the return value of `snprintf()` - it's not really a valid warning anyway, and increasing the size of static arrays like this can cause issues on platforms with limited stack space

DisasterMo commented:
Thank you both for your feedback.

Honestly, I do not understand what's so confusing or out of place about the index in non-playlists - personally I find it to be pure bliss to know the exact numeric position value of my most frequently used options (makes navigating RetroArch's settings jungle quite a bit easier, especially on XMB with its scrolling animation making the text hard to read). But I'm not here to argue. Have it your way, I say.

As for the thumbnail-text overlap: thank you, @jdgleaver, for reminding me of the vertical thumbnail disposition. I actually did not consider it during testing. While I would prefer the index display to have a fixed position and not jump around the screen, your suggested solution (b) seems the most reasonable - reducing the thumbnail size when so much screen estate is available on the left side of the screen is just too wasteful.

I'll see what I can do about the caching proposal.

jdgleaver commented:
> Honestly, I do not understand what's so confusing or out of place about the index in non-playlists - personally I find it to be pure bliss to know the exact numeric position value of my most frequently used options (makes navigating RetroArch's settings jungle quite a bit easier, especially on XMB with its scrolling animation making the text hard to read). But I'm not here to argue. Have it your way, I say.

Hmm... Well I don't strenuously object to having the index in all menu lists. The thing I don't like is the extra options overload - having two separate menu entries for controlling the visibility of this feature seems quite ugly to me. Also, `PLAYLIST_SHOW_ALL_IDX` is a misnomer - the setting is entirely unrelated to playlists :)

So I would say that you could keep the display turned on all the time, but:

- The `Show Playlist Entry Index` option should be hidden when using XMB
- Visibility should be controlled via a single `Show Entry Index` option, found under `User Interface > Appearance` only when using XMB

That is clear and concise, and should not confuse anyone!

DisasterMo commented:
> The thing I don't like is the extra options overload - having two separate menu entries for controlling the visibility of this feature seems quite ugly to me.

Oh, so you were explicitly objecting the extra option? Well then, I hope you will be glad to know that the inclusion of this option was only ever meant to be temporary, for testing purposes. I always had the intention to either seriously rework or outright remove it later - this here is a draft, after all.

I'll remove it for now, while I figure out how to make the index visible on the Music & Videos lists.

jdgleaver commented:
Ah, okay - sounds good :)

DisasterMo commented:
With this update:

- the enumeration string is shown in all lists internally recognised as 'playlist', as well as the Music & Videos lists and the lower levels of the Explore list (as for the 'load content' file browser, I do not know yet how to single out a 'sub-list' and don't feel like figuring it out now)
- the enumeration string will be drawn to the left of the selected entry if 'Vertical Thumbnail Disposition' is enabled (except for in the Explorer list)
- snprintf is only called if a change in entry number or list size has occurred (I am not sure how much better this solution is to before; I am not exactly intimately familiar with manual code optimisation techniques, so please help me out)

**Question:**
Should I edit the original description to reflect the most current state of the PR?
If so, should I replace it outright or just append '**UPDATE**' notes?

jdgleaver commented:
Thanks, nice updates :)

> the enumeration string is shown in all lists internally recognised as 'playlist', as well as the Music & Videos lists and the lower levels of the Explore list (as for the 'load content' file browser, I do not know yet how to single out a 'sub-list' and don't feel like figuring it out now)

You can use `xmb->is_file_list` to detect the `Load Content` file browser

> snprintf is only called if a change in entry number or list size has occurred (I am not sure how much better this solution is to before; I am not exactly intimately familiar with manual code optimisation techniques, so please help me out)

It would be better to populate the enumeration string inside `xmb_selection_pointer_changed()`, in this block here: https://github.com/libretro/RetroArch/blob/b1a3ea72c9f81450aa3ff9196b4376d48319d804/menu/drivers/xmb.c#L1433-L1502

That way you only have to perform those checks when the selection changes, rather than every frame.

> Should I edit the original description to reflect the most current state of the PR?
If so, should I replace it outright or just append 'UPDATE' notes?

You can leave the description as-is, if you like - we can track the updates via your comments :)

DisasterMo commented:
Whooo, boy! RetroArch needs a lot more comments, _badly_. 
- Moved entry index string population to `xmb_populate_entries()` (when changing lists) & `xmb_selection_pointer_changed()` (when switching vertically between entries).
- Fixed a bug in the original code due to which all sub-lists of certain `xmb->is_playlist` marked lists were also marked as such, which is nonsense. (Even though my 'solution' is a bit hacky, it _basically_ gives the wanted result and doesn't add too much overhead, so it's fine, surely...)
- Marked `Music` and `Videos` tabs and their 'deferred lists' (found under `Load Content` > `Playlists`) as `is_playlist`, because that seems reasonable.
- I decided against showing the entry index in the `Load Content` file explorer list. Unfortunately this seems to be the only list marked as `xmb->is_file_list`, and I want to make sure all file explorer like lists are consistent with each other. Which is... annoying to do, since there are so many of them with unique labels, see the `Directory` setting. x_x
- Removed the index from the `Explore` tab. Problem here is similar to the one I metioned fixing for `is_playlist`, but here it's a bit more tricky: all sub-lists litterally have the same `label`, for some reason, while having different depths, making distinguishing between them kinda hard.
**Can anyone point me to where those labels are set?** Because this _has_ to be changed.

jdgleaver commented:
Thanks for the update!

I'm afraid I'm utterly swamped by my backlog at the moment, so it will take me a little while to review this. I'll get on it as soon as I can! Please bear with me :)

jdgleaver commented:
Finally found some time to look at this - apologies for the delay!

> Whooo, boy! RetroArch needs a lot more comments, badly.

Indeed - this old legacy code is poorly documented... (it's getting better though, at least for new contributions...)

> Moved entry index string population to xmb_populate_entries() (when changing lists) & xmb_selection_pointer_changed() (when switching vertically between entries).

Very good!

> Fixed a bug in the original code due to which all sub-lists of certain xmb->is_playlist marked lists were also marked as such, which is nonsense. (Even though my 'solution' is a bit hacky, it basically gives the wanted result and doesn't add too much overhead, so it's fine, surely...)

Seems like a very sensible fix :)

> Marked Music and Videos tabs and their 'deferred lists' (found under Load Content > Playlists) as is_playlist, because that seems reasonable.

Good :)

> I decided against showing the entry index in the Load Content file explorer list. Unfortunately this seems to be the only list marked as xmb->is_file_list, and I want to make sure all file explorer like lists are consistent with each other. Which is... annoying to do, since there are so many of them with unique labels, see the Directory setting. x_x

I agree with you there. We don't really want to overcomplicate things....

> Removed the index from the Explore tab. Problem here is similar to the one I metioned fixing for xmb->is_playlist, but here it's a bit more tricky: all sub-lists litterally have the same label, for some reason, while having different depths, making distinguishing between them kinda hard.
Can anyone point me to where those labels are set? Because this has to be changed.

I agree with this decision too.

Unfortunately, you're not going to change those labels without a complete rewrite of the whole thing. I don't think anyone is up for that... :(

-----

So, I've checked the PR, and I like how it looks. This is a great addition :)

I only have a couple of comments:

- When vertical disposition is enabled, the index text must inherit its alpha value from the node label, otherwise you get 'leftovers' when scrolling horizontally through playlists
- When vertical disposition is enabled, I feel that the index text horizontal spacing is a little off - I would prefer there to be symmetry with the current item label spacing (also, that hard coded const in the y position bothers me...)

I have therefore prepared a tiny patch that fixes these issues, and cleans up a couple of other very minor things: 
[xmb.zip](https://github.com/libretro/RetroArch/files/6617923/xmb.zip)

Could you please apply this on-top of your PR, and check if things still work in the way you intended?

If it's okay, then we can go ahead and merge once you've squashed your commits :)

DisasterMo commented:
Thank you very much, @jdgleaver! Everything seems to be working as intended!

I would have really loved to see the index in the Explore lists as well, but if it would really require such a major rewrite... well maybe one day.

With this, I'm marking this PR as ready for merge!

jdgleaver commented:
Excellent, this is perfect now!

@twinaphex Please merge :)
